### PR TITLE
patch flang2 to work on windows

### PR DESCRIPTION
**Pull request by xoviat**
Also add linker directives option.

### CODE REVIEW

File: tools/flang2/flang2exe/CMakeLists.txt
Comment (bryanpkc) [inline tools/flang2/flang2exe/CMakeLists.txt:141]: Spurious empty line.

File: tools/flang2/flang2exe/CMakeLists.txt
Comment (bryanpkc) [inline tools/flang2/flang2exe/CMakeLists.txt:149]: Since flang2 is written in C++, it is probably not necessary to link libm explicitly.

File: tools/flang2/flang2exe/exputil.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/exputil.cpp:1111]: Why isn't the whole block deleted?

File: tools/flang2/flang2exe/kmpcutil.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/kmpcutil.cpp:535]: Why is this `assert` deleted?

File: tools/flang2/flang2exe/ll_structure.h
Comment (bryanpkc) [inline tools/flang2/flang2exe/ll_structure.h:723]: Please left-align this line with the others.

File: tools/flang2/flang2exe/llassem.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/llassem.cpp:196]: Spurious diff.

**Comment by xoviat** [comment]
@bryanpkc don't look at this yet. I was just waiting for an initial CI run.

File: tools/flang2/flang2exe/llassem.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/llassem.cpp:367]: Spurious diff. Same for the trailing spaces added to the lines below.

File: tools/flang2/flang2exe/llassem.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/llassem.cpp:1754]: Spurious diff.

File: tools/flang2/flang2exe/llassem.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/llassem.cpp:2608]: The coding style in Flang code base is K&R-like; please revert this diff.

File: tools/flang2/flang2exe/lldebug.cpp
Comment (bryanpkc) [inline tools/flang2/flang2exe/lldebug.cpp:1788]: Please don't add the blank lines. The original code was fine.

File: tools/flang2/flang2exe/kmpcutil.cpp
Comment (xoviat) [reply tools/flang2/flang2exe/kmpcutil.cpp:535]: it's not true on windows

File: tools/flang2/flang2exe/outliner.cpp
Comment (pawosm-arm) [inline tools/flang2/flang2exe/outliner.cpp]: Windows specific header! This is fixed couple of commits later. Can you squash your commits into one to avoid further confusion?

**Review by pawosm-arm — CHANGES_REQUESTED** [review]
I'm not an Windows expert, quite contrary, been avoiding it for last 20+ years, so I can only address the general layout of this PR which is made of several commits touching overlapping pieces of code which confuses potential reviewers.

**Review by pawosm-arm — APPROVED** [review]

File: tools/flang2/flang2exe/symacc.cpp
Comment (kaadam) [inline tools/flang2/flang2exe/symacc.cpp:44]: Is it necessary comment it out if it is guarded?

Comment (xoviat) [reply tools/flang2/flang2exe/symacc.cpp:44]: I seem to recall that it failed the CI.

Comment (xoviat) [reply tools/flang2/flang2exe/symacc.cpp:44]: https://github.com/flang-compiler/flang/runs/2282723391#step:10:387

Comment (xoviat) [reply tools/flang2/flang2exe/symacc.cpp:44]: Maybe this block should be removed.

**Comment by xoviat** [comment]
cc @kiranchandramohan

**Comment by kiranchandramohan** [comment]
@kaadam is your review complete? If so, can you approve.
I will get to this on Saturday.

File: tools/flang2/flang2exe/symacc.cpp
Comment (bryanpkc) [reply tools/flang2/flang2exe/symacc.cpp:44]: Is it not possible to compare with a different size when building on Windows? e.g.
```c
#ifdef _WIN64
  assert(sizeof_SYM == 48, "bad SYM size", sizeof_SYM, ERR_Fatal);
#else
  assert(sizeof_SYM == 36, "bad SYM size", sizeof_SYM, ERR_Fatal);
#endif
```

Comment (kiranchandramohan) [reply tools/flang2/flang2exe/symacc.cpp:44]: Why have you changed from ERR_Fatal to its value 4? If you keep it as ERR_Fatal wouldn't it work?

The patch looks OK to me.

File: tools/flang2/flang2exe/CMakeLists.txt
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/CMakeLists.txt:95]: Could you remind me why we need this change?

**Review by kiranchandramohan — COMMENTED** [review]
@xoviat  Looks Good to Me. Can you address the removal of the assert?

File: tools/flang2/flang2exe/CMakeLists.txt
Comment (xoviat) [reply tools/flang2/flang2exe/CMakeLists.txt:95]: There's only one existing target architecture.

File: tools/flang2/flang2exe/symacc.cpp
Comment (kaadam) [reply tools/flang2/flang2exe/symacc.cpp:44]: 64bit Windows uses the LLP64 data model where i.e. long is still 32bit. SYM size as fas as I remember should be 44. These sym size comes from legacy_ints definitons, am I right?

**Review by kaadam — APPROVED** [review]
Sorry for the delay, I forgot to add lgtm.

**Comment by xoviat** [comment]
perhaps `-Linux` should be removed?

image

**Comment by xoviat** [comment]
Perhaps `-Linux` should be removed?

image

**Comment by xoviat** [comment]
@kiranchandramohan Anything else?

File: tools/flang2/flang2exe/CMakeLists.txt
Comment (isuruf) [reply tools/flang2/flang2exe/CMakeLists.txt:95]: https://github.com/flang-compiler/flang/pull/726 fixes this

File: tools/flang2/flang2exe/symacc.cpp
Comment (kiranchandramohan) [reply tools/flang2/flang2exe/symacc.cpp]: @xoviat a reply to @kaadam's and @bryanpkc's point here will be great.

Comment (kiranchandramohan) [reply tools/flang2/flang2exe/symacc.cpp]: ping @xoviat

File: tools/flang2/flang2exe/symacc.cpp
Comment (kiranchandramohan) [reply tools/flang2/flang2exe/symacc.cpp:44]: Ping @kaadam

**Comment by xoviat** [comment]
Actually, let's skip the sym size on windows for now and merge this, since this won't break unix. I'll open an issue to addres this.

**Comment by xoviat** [comment]
@kaadam You can PR your changes to `master` after this is merged.

**Comment by xoviat** [comment]
ping @kiranchandramohan

File: tools/flang2/flang2exe/CMakeLists.txt
Comment (bryanpkc) [inline tools/flang2/flang2exe/CMakeLists.txt:170]: @xoviat Why was this line ever needed? It makes the destination directory different for `flang1` and `flang2`, and seems to have caused a build issue for some users (#1338).

**Pull request merged**
